File: placementprepration/stack.cpp
#include<iostream>
using namespace std;

class Stack {
private:
    int top;
    int maxSize;
    int* array;

public:
    Stack(int size) {
        maxSize = size;
        array = new int[maxSize];
        top = -1;
    }

    bool isEmpty() {
        return (top == -1);
    }

    bool isFull() {
        return (top == maxSize - 1);
    }

    void push(int value) {
        if (isFull()) {
            cout << "Stack Overflow! Cannot push " << value << endl;
            return;
        }
        array[++top] = value;
        cout << value << " pushed to stack" << endl;
    }

    int pop() {
        if (isEmpty()) {
            cout << "Stack Underflow! Cannot pop from empty stack" << endl;
            return -1;
        }
        int value = array[top--];
        return value;
    }

    int peek() {
        if (isEmpty()) {
            cout << "Stack is empty!" << endl;
            return -1;
        }
        return array[top];
    }

    ~Stack() {
        delete[] array;
    }
};

int main() {
    Stack s(5);  
    s.push(10);
    s.push(20);
    s.push(30);
    
    cout << "Top element is: " << s.peek() << endl;
    
    cout << s.pop() << " popped from stack" << endl;
    cout << s.pop() << " popped from stack" << endl;
    
    cout << "Is stack empty? " << (s.isEmpty() ? "Yes" : "No") << endl;
    
    return 0;
}

File: placementprepration/stack.py
class Stack:
    def __init__(self, size):
        self.max_size = size
        self.array = [None] * size
        self.top = -1

    def is_empty(self):
        return self.top == -1

    def is_full(self):
        return self.top == self.max_size - 1

    def push(self, value):
        if self.is_full():
            return None
        self.top += 1
        self.array[self.top] = value
        return value

    def pop(self):
        if self.is_empty():
            return None
        value = self.array[self.top]
        self.top -= 1
        return value

    def peek(self):
        if self.is_empty():
            return None
        return self.array[self.top]


if __name__ == "__main__":
    s = Stack(5)
    s.push(10)
    s.push(20)
    s.push(30)
    print("Top element is:", s.peek())
    print(s.pop(), "popped from stack")
    print(s.pop(), "popped from stack")
    print("Is stack empty?", "Yes" if s.is_empty() else "No")
